### ISSUE TITLE
Broken pipe when copying in tree mode

### ISSUE THREAD

**Issue by degremont** [cea-hpc/clustershell#308]
In tree mode, when copying a file on local and remote nodes, a broken pipe error could happened (racy, based on nodeset size).
``clush: [Errno 32] Broken pipe``

WorkerTree keeps a list of worker it has spawned. Those workers will likely be gateways but could also be WorkerSsh for all directly connected nodes.
When write is used this is propagated to all these workers in ``WorkerTree.write()``

What happens if WorkerSsh manages ScpClient? WorkerTree will try to write on those EngineClients.
Calling ``write()`` on copyclient instances does not really make sense.

Sometimes, this ``write()`` call will fail with Broken Pipe error. Likely because the scp process is already dead. (I'm still wondering how we could try to write on an already closed process)

How to reproduce:
* Use a topology with enough directly attached nodes, and few nodes behind a gateway 
* Copy a file to those nodes

File size and node responsiveness should be well tuned to reproduce this race where ``write()`` is send to already finished scp instances.

**Comment by degremont** [comment]
Broken pipe could also be triggered using simpler commands:

``$ cat /tmp/my230mb_file | clush -w directnode,remotenode true``

Sometimes (racy), this could raise IOError (broken pipe), sometime this raises 

```
File "ClusterShell/Propagation.py", line 312, in write
   assert id(worker) in self.workers
```

**Comment by thiell** [comment]
This is related to #196. I might be able to find a quick fix for 1.7.2.

My understanding is that the data from stdin don't have time to be written to the remote processes, indeed they are closing too quickly, thus generating broken pipe errors. A common scenario is:
* WorkerTree spawns gateways and regular ssh workers, but the ssh (direct) workers are much faster than the gateways stuff, so close earlier and generate a broken pipe error on WorkerTree.write() [called from the stdin thread] (so that closes everything: exception -> cleaning -> _close (abort))
* then the gateway itself got broken pipe, but necessarily in the same way, but because the parent closed too early due to the last point: ev_close() is called because the process exits -> ev_timer(None) is called to finalize grooming and tries to write() back to the parent -> EPIPE.

Thinking about it, this is normal to have broken pipe errors on write() calls, it's not a engine failure and can happen at any time. In both case, we could just ignore broken pipe errors on write() as the engine should receive a EPOLLERR event shortly afterwards and unregister the associated FD. ssh just seems to ignore them in that case. I have to do some tests just to be 100% sure, but that would mean it's a very easy fix.

Re: write on scp processes, that's almost a separate issue. I will start to fix the issue reported in your first comment.

**Comment by degremont** [comment]
You're right, my second comment is a duplicate of #196.

But the original purpose of this ticket is ``clush --copy`` failing in tree mode. This is what needs to be fixed here. Let's use #196 for the other problem.

**Comment by thiell** [comment]
Ok, ideally both issues should be fixed in 1.7.2, I will work on patches.

**Comment by thiell** [comment]
Patch proposal here: https://review.gerrithub.io/277525
The `WorkerTree.gwtargets` dict wasn't properly cleaned, so in case of massive write() on WorkerTree, after a gateway has finished, this would silently relaunch a new gateway with no targets, leading to unexpected results... :)
I don't seem to be able to see the problem anymore. Of course, better tests of the tree mode would be preferred, thanks for opening #310 for that.
